### PR TITLE
[extractor/nbc] Add NBCStationsIE

### LINKED ISSUE #4571: Add NBC TV News Sites,  www.NBCLosAngeles.com , www.NBCBayArea.com, www.NBCSanDiego.com ..etc
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.07.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

_No response_

### Example URLs

https://www.nbclosangeles.com/news/local/junior-lifeguards-pull-pilot-from-small-plane-crash/2947359/

https://www.nbcbayarea.com/news/national-international/small-banner-plane-crash-huntington-beach/2955788/

https://www.nbcsandiego.com/live
https://www.nbcconnecticut.com/
https://www.nbcwashington.com/
https://www.nbcmiami.com/
https://www.nbcchicago.com/
https://www.nbcboston.com/
https://www.nbcnewyork.com/
https://www.nbcphiladelphia.com/
https://www.telemundopr.com/
https://www.necn.com/

List from: https://en.wikipedia.org/wiki/NBC_Owned_Television_Stations

### Provide a description that is worded well enough to be understood

Please add support for all of the mentioned domains in the example links.

The sites seem to have a very similar if not the same exact backend, just different end-user content for different region. They also link to their other sister sites a lot, for example the news article on nbcsandiego.com has an embedded video that points to nbclosangeles.com

Sites seem to use /master.m3u8, but I got it from Firefox's "network" tab and not from page source.

A Livestream for each site are accessible on the domain/live address for all sites. However livestreams are either geographically restricted or locked under subscribtion. The geographic restriction that I am seeing is most likely due to EU GDPR laws which many USA based news sites were not willing to support so many of them completely banned all EU IP addresses for the whole domain. Larger establishments have since started supporting EU GDPR laws to allow EU citizen to access the sites, which includes these NBC local TV stations, with the exception of livestreams it seems.

www.nbcbayarea.com/live gives me 
> Login with your TV provider to watch unlimited NBC content on
> [www.nbc.com/live]()

www.nbconnecticut.com/live gives a black unseen message with just a close button, inspecting webpage source gives: 

> <div class="tpErrorOverlay tpPlayerArea" style="">
>       <div class="tpErrorCard tpCard">
>         <div class="tpErrorTitle" tabindex="0"><span>Geographic Restriction</span></div>
>         <div class="tpErrorBody" tabindex="0"><span>This content is not available in your location.</span></div>
>       </div>
>       <div class="tpActions">
> <button class="tpCloseButton" value="close" aria-label="Close">Close</button>
> </div>
>     </div>

TOS:

> To the extent that NBCUniversal makes Content available to you via a widget, embedded player or other technology that allows you to embed or stream Content on or to another site, you may not modify, enhance, remove, interfere with, or otherwise alter in any way any portion of such technology, any digital rights management mechanism, device, or other content protection or access control measure associated with Content, or any advertisement delivered with Content. You are also prohibited from disabling, modifying, interfering with or otherwise circumventing any technology to allow users to view Content without: (i) displaying visibly both Content and all surrounding elements (including the graphical user interface, any advertising, copyright notices, and trademarks); and (ii) having full access to all functionality permitting viewing of Content, including, without limitation, all video quality and display functionality and all interactive, elective or click-through advertising functionality.

For the first sentence: The videos/media in question is publicly available and has no DRM.
Second sentence: yt-dlp and it's users would not be serving the Content to other users.

So I think it's fine for a personal copy then, right?

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vvU', 'https://www.nbclosangeles.com/news/local/junior-lifeguards-pull-pilot-from-small-plane-crash/2947359/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (zip)
[debug] Python 3.10.5 (CPython 64bit) - Linux-5.18.12-1-default-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.1 (fdk,setts), ffprobe 5.1
[debug] Optional libraries: brotli-1.0.9, certifi-2021.10.08, sqlite3-2.6.0, websockets-10.1
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.07.18, Current version: 2022.07.18
yt-dlp is up to date (2022.07.18)
[debug] [generic] Extracting URL: https://www.nbclosangeles.com/news/local/junior-lifeguards-pull-pilot-from-small-plane-crash/2947359/
[generic] 2947359: Downloading webpage
WARNING: [generic] Falling back on generic information extractor.
[generic] 2947359: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://www.nbclosangeles.com/news/local/junior-lifeguards-pull-pilot-from-small-plane-crash/2947359/
Traceback (most recent call last):
  File "/usr/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1431, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1506, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/bin/yt-dlp/yt_dlp/extractor/common.py", line 644, in extract
    ie_result = self._real_extract(url)
  File "/usr/bin/yt-dlp/yt_dlp/extractor/generic.py", line 4108, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.nbclosangeles.com/news/local/junior-lifeguards-pull-pilot-from-small-plane-crash/2947359/

------------------------------

[debug] Command-line config: ['-vU', '-F', 'https://nbclim-f.akamaihd.net/i/Prod/NBCU_LM_VMS_-_KNBC/599/827/17552149962-,1080p,720p,540phigh,540plow,360p,234p,nbcstations.mp4.csmil/master.m3u8']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (zip)
[debug] Python 3.10.5 (CPython 64bit) - Linux-5.18.12-1-default-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.1 (fdk,setts), ffprobe 5.1
[debug] Optional libraries: brotli-1.0.9, certifi-2021.10.08, sqlite3-2.6.0, websockets-10.1
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.07.18, Current version: 2022.07.18
yt-dlp is up to date (2022.07.18)
[debug] [generic] Extracting URL: https://nbclim-f.akamaihd.net/i/Prod/NBCU_LM_VMS_-_KNBC/599/827/17552149962-,1080p,720p,540phigh,540plow,360p,234p,nbcstations.mp4.csmil/master.m3u8
[generic] master: Downloading webpage
[debug] Identified a direct video link
[generic] master: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for master:
ID   EXT RESOLUTION │   TBR PROTO │ VCODEC       VBR ACODEC    ABR
──────────────────────────────────────────────────────────────────
304  mp4 416x234    │  304k m3u8  │ avc1.77.30  304k mp4a.40.2  0k
820  mp4 640x360    │  820k m3u8  │ avc1.77.30  820k mp4a.40.2  0k
1535 mp4 960x540    │ 1535k m3u8  │ avc1.77.30 1535k mp4a.40.2  0k
1905 mp4 960x540    │ 1905k m3u8  │ avc1.77.30 1905k mp4a.40.2  0k
2742 mp4 1280x720   │ 2742k m3u8  │ avc1.77.30 2742k mp4a.40.2  0k
4103 mp4 1920x1080  │ 4103k m3u8  │ avc1.77.30 4103k mp4a.40.2  0k
```

### PR DESCRIPTION
**Pull request by bashonly**
</details>

Adds support for NBC-owned television station news sites (referred to as `nbcstations` internally). These sites use `theplatform` a bit differently than the other extractors, so it was not beneficial to make this IE a subclass of `ThePlatformIE`.

Fixes #4571

Supports extraction of VOD and livestreams -- but NOT the live TV streams, ONLY the newscast livestreams that are available only during the actual broadcasts. The Live TV streams are region-locked and it would be a fool's errand to try to implement them, IMO.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/nbc.py [hunk 1]
```diff
@@ -584,3 +589,172 @@ def _real_extract(self, url):
...
+            f'https://link.theplatform.com/s/{pdk_acct}/{player_id}', video_id,
+            headers=headers, query=query, note='Downloading manifest')
+        if live:
+            manifest_url = self._search_regex(r'<video src="([^"]*)', manifest, 'manifest URL')
+        else:
+            manifest_url = urlh.geturl()
+
+        formats.extend(self._extract_m3u8_formats(
+            manifest_url, video_id, 'mp4', entry_protocol=entry_protocol, headers=headers,
+            m3u8_id='hls', fatal=live, live=live, errnote='No HLS formats found'))
```
[the remark this answers is not in the document]
Comment (bashonly) [reply yt_dlp/extractor/nbc.py:748]: The variable `live` is `True` if the video is a livestream.  Livestreams only have HLS formats so then we want this to be fatal, whereas VOD can also have http mp4 formats so then we want this to be non-fatal

File: yt_dlp/extractor/nbc.py [hunk 2]
```diff
@@ -584,3 +590,170 @@ def _real_extract(self, url):
...
+
+
+class NBCStationsIE(InfoExtractor):
+    _VALID_URL = r'''(?x)
+                         https?://(?:www\.)?(?P<site>
+                             nbcbayarea|nbcboston|nbcchicago|nbcconnecticut|nbcdfw|nbclosangeles|
+                             nbcmiami|nbcnewyork|nbcphiladelphia|nbcsandiego|nbcwashington|necn|
+                             telemundo52|telemundoarizona|telemundochicago|telemundonuevainglaterra
+                         )\.com/(?:[^/]+/)*(?P<id>[^/?]+)/?(?:$|[#?])
+                    '''
```
Comment (pukkandan) [inline yt_dlp/extractor/nbc.py:602]:
```suggestion
    _DOMAIN_RE = '|'.join(map(re.escape, (
        'nbcbayarea', 'nbcboston', 'nbcchicago', 'nbcconnecticut', 'nbcdfw', 'nbclosangeles',
        'nbcmiami', 'nbcnewyork', 'nbcphiladelphia', 'nbcsandiego', 'nbcwashington',
        'necn', 'telemundo52', 'telemundoarizona', 'telemundochicago', 'telemundonuevainglaterra',
    )))
    _VALID_URL = rf'https?://(?:www\.)?(?P<site>{_DOMAIN_RE})\.com/(?:[^/?#]+/)*(?P<id>[^/?#]+)/?(?:$|[#?])'
```

File: yt_dlp/extractor/nbc.py [hunk 3]
```diff
@@ -584,3 +590,170 @@ def _real_extract(self, url):
...
+            r'<script>var\s*nbc\s*=\s*', webpage, 'NBC JSON data', video_id)
+        pdk_acct = nbc_data.get('pdkAcct') or 'Yh1nAC'
+        fw_ssid = traverse_obj(nbc_data, ('video', 'fwSSID'))
+        fw_network_id = traverse_obj(nbc_data, ('video', 'fwNetworkID'), default='382114')
+
+        video_data = self._parse_json(self._html_search_regex(
+            r'data-videos="([^"]*)"', webpage, 'video data', fatal=False, default='{}'), video_id)
+        video_data = variadic(video_data)[0]
+        video_data.update(self._parse_json(self._html_search_regex(
+            r'data-meta="([^"]*)"', webpage, 'metadata', fatal=False, default='{}'), video_id))
```
Comment (pukkandan) [inline yt_dlp/extractor/nbc.py:656]:
```suggestion
            r'data-videos="([^"]*)"', webpage, 'video data', fdefault='{}'), video_id)
        video_data = variadic(video_data)[0]
        video_data.update(self._parse_json(self._html_search_regex(
            r'data-meta="([^"]*)"', webpage, 'metadata', default='{}'), video_id))
```
default implies non-fatal

**Review by pukkandan — APPROVED** [review]

**Comment by antdude** [comment]
Is this issue still under development? I still see this problem in two different Linux PCs:

Hmm, it doesn't work for me in two different Linux boxes:

[debug] Command-line config: ['-U', '-v', 'https://www.nbclosangeles.com/news/local/bear-goes-for-a-swim-in-monrovia-home-during-rain/3073958/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2023.01.06 [6becd25] (zip)
[debug] Python 3.11.1 (CPython x86_64 64bit) - Linux-6.0.14-300.fc37.x86_64-x86_64-with-glibc2.36 (OpenSSL 3.0.5 5 Jul 2022, glibc 2.36)
[debug] exe versions: ffmpeg 5.1.2 (setts), ffprobe 5.1.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1760 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2023.01.06, Current version: 2023.01.06
yt-dlp is up to date (2023.01.06)
[NBCStations] Extracting URL: https://www.nbclosangeles.com/news/local/bear-goes-for-a-swim-in-monrovia-home-during-rain/3073958/
[NBCStations] 3073958: Downloading webpage
[NBCStations] 3073958: Downloading manifest
[NBCStations] 3073958: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
ERROR: encoding with 'idna' codec failed (UnicodeError: Invalid character '�')
Traceback (most recent call last):
  File "/usr/lib64/python3.11/encodings/idna.py", line 179, in encode
    result.extend(ToASCII(label))
                  ^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/encodings/idna.py", line 74, in ToASCII
    label = nameprep(label)
            ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/encodings/idna.py", line 38, in nameprep
    raise UnicodeError("Invalid character %r" % c)
UnicodeError: Invalid character '�'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/YoutubeDL.py", line 1502, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/YoutubeDL.py", line 1599, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/YoutubeDL.py", line 1658, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/YoutubeDL.py", line 2662, in process_video_result
    format['http_headers'] = self._calc_headers(full_format_info)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/YoutubeDL.py", line 2355, in _calc_headers
    cookies = self._calc_cookies(info_dict['url'])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/YoutubeDL.py", line 2367, in _calc_cookies
    pr = sanitized_Request(url)
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/utils.py", line 772, in sanitized_Request
    url, auth_header = extract_basic_auth(escape_url(sanitize_url(url)))
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ant/bin/videoDLers/yt-dlp/yt_dlp/utils.py", line 3120, in escape_url
    netloc=url_parsed.netloc.encode('idna').decode('ascii'),
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeError: encoding with 'idna' codec failed (UnicodeError: Invalid character '�')


[debug] Command-line config: ['-U', '-v', 'https://www.nbclosangeles.com/news/local/bear-goes-for-a-swim-in-monrovia-home-during-rain/3073958/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2023.01.06 [6becd25] (zip)
[debug] Python 3.9.2 (CPython x86_64 64bit) - Linux-5.10.0-20-amd64-x86_64-with-glibc2.31 (OpenSSL 1.1.1n  15 Mar 2022, glibc 2.31)
[debug] exe versions: ffmpeg 4.3.5-0, ffprobe 4.3.5-0, rtmpdump 2.4
[debug] Optional libraries: certifi-2020.06.20, pyxattr-0.7.2, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1760 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2023.01.06, Current version: 2023.01.06
yt-dlp is up to date (2023.01.06)
[NBCStations] Extracting URL: https://www.nbclosangeles.com/news/local/bear-goes-for-a-swim-in-monrovia-home-during-rain/3073958/
[NBCStations] 3073958: Downloading webpage
[NBCStations] 3073958: Downloading manifest
[NBCStations] 3073958: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
ERROR: encoding with 'idna' codec failed (UnicodeError: Invalid character '�')
Traceback (most recent call last):
  File "/usr/lib/python3.9/encodings/idna.py", line 181, in encode
    result.extend(ToASCII(label))
  File "/usr/lib/python3.9/encodings/idna.py", line 76, in ToASCII
    label = nameprep(label)
  File "/usr/lib/python3.9/encodings/idna.py", line 38, in nameprep
    raise UnicodeError("Invalid character %r" % c)
UnicodeError: Invalid character '�'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ant/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1502, in wrapper
    return func(self, *args, **kwargs)
  File "/home/ant/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1599, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/home/ant/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1658, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "/home/ant/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 2662, in process_video_result
    format['http_headers'] = self._calc_headers(full_format_info)
  File "/home/ant/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 2355, in _calc_headers
    cookies = self._calc_cookies(info_dict['url'])
  File "/home/ant/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 2367, in _calc_cookies
    pr = sanitized_Request(url)
  File "/home/ant/bin/yt-dlp/yt_dlp/utils.py", line 772, in sanitized_Request
    url, auth_header = extract_basic_auth(escape_url(sanitize_url(url)))
  File "/home/ant/bin/yt-dlp/yt_dlp/utils.py", line 3120, in escape_url
    netloc=url_parsed.netloc.encode('idna').decode('ascii'),
UnicodeError: encoding with 'idna' codec failed (UnicodeError: Invalid character '�')\

**Pull request merged**